### PR TITLE
Fix `make cookbook` when certain cookbooks are guarded

### PR DESCRIPTION
**Pull request by geektoni**
This should fix issue #3520. These changes fix also ``` make cookbook_sphinx_linkcheck```.

The command ```make cookbook``` still gives me warnings, though, but I don't understand if they are caused by my changes or if they were already there.
See warnings here: http://pastebin.com/chxfGvaq

### CODE REVIEW

**Comment by geektoni** [comment]
Mmmh, I think you are pretty right. It was more like a "quick fix" than a proper solution. I refactored the code a bit to make it more readable and modular. This new approach should be better.

**<off_topic>**
Can I ask why we have to deal with missing dependencies? We could simply add these libraries as required to compile shogun, instead of setting up these "workarounds". Obviously, I'm sure that you have a bunch of good reasons for doing so. ;)
**</off_topic>**

**Comment by karlnapf** [comment]
Yes, an overwhealming number of good reasons for that :)
Come to IRC and we can discuss

File: cmake/FindMetaExamples.cmake [hunk 1]
```diff
@@ -15,40 +15,75 @@ function(is_set CONFIG_FILE FLAG)
...
+    is_set(${SHOGUN_CONFIG} HAVE_LAPACK)
+    is_set(${SHOGUN_CONFIG} USE_SVMLIGHT)
 
-	# temporary hacks to exclude certain meta examples that have dependencies
-	IF(NOT HAVE_NLOPT)
-		LIST(REMOVE_ITEM META_EXAMPLE_LISTINGS
-			${CMAKE_SOURCE_DIR}/examples/meta/src/gaussian_processes/gaussian_process_regression.sg)
+    IF(NOT HAVE_NLOPT)
+		LIST(APPEND EXCLUDED_META_EXAMPLES
+			gaussian_processes/gaussian_process_regression.sg)
```
Comment (karlnapf) [inline cmake/FindMetaExamples.cmake]: +1 this is cleaner!

File: cmake/FindMetaExamples.cmake [hunk 2]
```diff
@@ -15,40 +15,75 @@ function(is_set CONFIG_FILE FLAG)
...
+
+endfunction()
+
+# Remove meta example that cannot be built because of missing dependencies
+function(find_meta_examples)
+
+	FILE(GLOB_RECURSE META_EXAMPLE_LISTINGS ${CMAKE_SOURCE_DIR}/examples/meta/src/*.sg)
+    get_excluded_meta_examples()
+
+	# temporary hacks to exclude certain meta examples that have dependencies
```
Comment (karlnapf) [inline cmake/FindMetaExamples.cmake]: I guess this comment can be removed now, as it is a bit nicer

File: cmake/FindMetaExamples.cmake [hunk 3]
```diff
@@ -15,40 +15,75 @@ function(is_set CONFIG_FILE FLAG)
...
+    FOREACH(META_EXAMPLE ${EXCLUDED_META_EXAMPLES})
+        LIST(REMOVE_ITEM META_EXAMPLE_LISTINGS ${CMAKE_SOURCE_DIR}/examples/meta/src/${META_EXAMPLE})
+    ENDFOREACH()
+
 	SET(META_EXAMPLES ${META_EXAMPLE_LISTINGS} PARENT_SCOPE)
 endfunction()
+
+# Get the cookbook pages we want to exclude from build
+function(find_excluded_meta_examples)
```
Comment (karlnapf) [inline cmake/FindMetaExamples.cmake]: I think this function should be named differently, as it is about cookbook pages, not meta examples themselves

File: cmake/FindMetaExamples.cmake [hunk 4]
```diff
@@ -15,40 +15,75 @@ function(is_set CONFIG_FILE FLAG)
...
+
+    FOREACH(META_EXAMPLE ${EXCLUDED_META_EXAMPLES})
+        LIST(APPEND EXCLUDED_COOKBOOK_PAGES examples/${META_EXAMPLE})
+    ENDFOREACH()
+
+    # Generate a string with all the meta examples separated by commas.
+    # This is made since Sphinx's exclude_patterns option wants
+    # the list's items separated by commas, but cmake's lists use
+    # semicolons instead.
+    # See: https://cmake.org/cmake/help/v3.3/command/list.html
```
Comment (karlnapf) [inline cmake/FindMetaExamples.cmake]: good to put that comment

File: doc/cookbook/CMakeLists.txt [hunk 5]
```diff
@@ -1,5 +1,7 @@
 #PROJECT(sphinxdoc)
 #cmake_minimum_required(VERSION 2.8.8)
+include(FindMetaExamples)
+find_excluded_meta_examples()
```
Comment (karlnapf) [inline doc/cookbook/CMakeLists.txt]: rename as said above

File: cmake/FindMetaExamples.cmake [hunk 6]
```diff
@@ -15,40 +15,75 @@ function(is_set CONFIG_FILE FLAG)
...
 	SET(META_EXAMPLES ${META_EXAMPLE_LISTINGS} PARENT_SCOPE)
 endfunction()
+
+# Get the cookbook pages we want to exclude from build
+function(find_excluded_meta_examples)
+
+    get_excluded_meta_examples()
+
+    FOREACH(META_EXAMPLE ${EXCLUDED_META_EXAMPLES})
+        LIST(APPEND EXCLUDED_COOKBOOK_PAGES examples/${META_EXAMPLE})
```
Comment (karlnapf) [inline cmake/FindMetaExamples.cmake]: one thing to consider here: Some meta examples do not have cookbook pages. So they do not need to be excluded technically. Not sure whether that causes problems, but it is definitely cleaner to only work on the existing cookbook pages

**Comment by geektoni** [comment]
I made the suggested changes @karlnapf. 
By the way, I would definitely like to know more about these management choices about shogun dependencies. Are you shogun developers "always online" on IRC or there is a preferred time frame?

**Comment by karlnapf** [comment]
Fine to merge from my side. @vigsterkr ?

**Comment by karlnapf** [comment]
@geektoni IRC hours depend. I am UTC 11am-7pm usually. Others might differ.

**Comment by vigsterkr** [comment]
go

**Comment by karlnapf** [comment]
Weird failures in travis, I just restarted to be safe.

**Pull request merged**